HW9/AVLTree/AVLTree.c

```diff
@@ -0,0 +1,270 @@
+﻿#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#include "AVLTree.h"
+
+typedef struct Dictionary {
+    const char* value;
+    const char* key;
+    Dictionary* left;
+    Dictionary* right;
+    int height;
+} Dictionary;
+
+int getHeight(Dictionary* node) {
+    return node ? node->height : -1;
+}
+
+void updateHeight(Dictionary* node) {
+    if (node != NULL) {
+        node->height = 1 + (getHeight(node->left) > getHeight(node->right) ? getHeight(node->left) : getHeight(node->right));
+    }
+}
+
+int getBalance(Dictionary* node) {
+    return node ? getHeight(node->left) - getHeight(node->right) : 0;
+}
+
+Dictionary* createDictionary(const char* value, const char* key) {
+    Dictionary* newNode = malloc(sizeof(Dictionary));
+    if (newNode == NULL) {
+        return NULL;
+    }
+
+    newNode->value = strdup(value);
+    if (newNode->value == NULL) {
+        free(newNode);
+        return NULL;
+    }
+
+    newNode->key = strdup(key);
+    if (newNode->key == NULL) {
+        free((char*)newNode->value);
+        free(newNode);
+        return NULL;
+    }
+
+    newNode->height = 0;
+    newNode->left = NULL;
+    newNode->right = NULL;
+    return newNode;
+}
+
+Dictionary* rotateLeft(Dictionary* node) {
+    Dictionary* rightChild = node->right;
+    Dictionary* leftOfRight = rightChild->left;
+
+    rightChild->left = node;
+    node->right = leftOfRight;
+
+    updateHeight(node);
+    updateHeight(rightChild);
+    return rightChild;
+}
+
+Dictionary* rotateRight(Dictionary* node) {
+    Dictionary* leftChild = node->left;
+    Dictionary* rightOfLeft = leftChild->right;
+
+    leftChild->right = node;
+    node->left = rightOfLeft;
+
+    updateHeight(node);
+    updateHeight(leftChild);
+    return leftChild;
+}
+
+Dictionary* bigLeftRotate(Dictionary* node) {
+    node->right = rotateRight(node->right);
+    return rotateLeft(node);
+}
+
+Dictionary* bigRightRotate(Dictionary* node) {
+    node->left = rotateLeft(node->left);
+    return rotateRight(node);
+}
+
+Dictionary* add(Dictionary* node, const char* key, const char* value) {
+    if (key == NULL || value == NULL) {
+        return NULL;
+    }
+    if (node == NULL) {
+        return createDictionary(value, key);
+    }
+
+    int tmp = strcmp(key, node->key);
+
+    if (tmp < 0) {
+        node->left = add(node->left, key, value);
+        if (node->left == NULL) {
+            return NULL;
+        }
+    }
+    else if (tmp > 0) {
+        node->right = add(node->right, key, value);
+        if (node->right == NULL) {
+            return NULL;
+        }
+    }
+    else {
+        char* temp = strdup(value);
+        if (temp == NULL) {
+            return NULL;
+        }
+        free((char*)node->value);
+        node->value = strdup(value);
```

Comment:
```suggestion
        node->value = tmp;
```
Провал :) Не воспринимайте мои подсказки столь буквально